components/molecule/modal/src/HoC/WithUrlState.js

```diff
@@ -0,0 +1,69 @@
+import React, {useEffect, useState} from 'react'
+import PropTypes from 'prop-types'
+
+const getUrl = ({hash: hashName} = {}) => {
+  const location = window.location
+  const hash = hashName ? `#${hashName}` : ''
+  return `${location.origin}${location.pathname}${location.search}${hash}`
+}
+
+const checkHash = hash => window.location.hash.includes(hash)
+
+const emptyFn = () => {}
+
+/**
+ * It must be refactored with react-router hooks (in React Router > 5)
+ * - useHistory
+ * - useLocation
+ */
+export default BaseComponent => {
+  const displayName = BaseComponent.displayName
+
+  if (typeof window === 'undefined') return BaseComponent
+
+  const WithUrlState = props => {
+    const {hash, openModalTrigger, onClose, ...rest} = props
+    const {isOpen} = rest
+    const [isPopStateChange, setIsPopStateChange] = useState(checkHash(hash))
+
+    const _handlePopState = ev => {
+      const hasHash = checkHash(hash)
+      setIsPopStateChange(true)
+
+      if (hasHash) openModalTrigger()
+      else onClose()
+    }
+
+    const _handleStateChange = () => {
+      if (isPopStateChange) return setIsPopStateChange(false)
+
+      const nextUrl = isOpen ? getUrl({hash}) : getUrl()
+      window.history.pushState({hash}, '', nextUrl)
+    }
+
+    useEffect(() => {
+      _handleStateChange()
+    }, [hash, isOpen]) // eslint-disable-line
+
+    useEffect(() => {
+      if (isPopStateChange) openModalTrigger()
+
+      window.onpopstate = _handlePopState
+      return () => {
+        window.onpopstate = emptyFn
+      }
+    }, []) // eslint-disable-line
```

Comment:
Also, instead of not placing `isPopStateChange` as a dependency, we may just use `hash` and do something like:
```js
if (checkHash(hash)) openModalTrigger()
```

That way `eslint-disable-line` comment becomes avoidable, which is desired.

Reply:
It must remains because hash not only changes when popState event triggered.
`_handleStateChange` should not modify history if hash changes only in popState (when user click next/prev arrows in browser)